### PR TITLE
Replace PBKDF2 with Argon2id

### PR DESCRIPTION
**Pull request by umutcamliyurt**
Build and test it before merging

### CODE REVIEW

**Comment by umutcamliyurt** [comment]
If it doesn't work on your system i will use hardcoded salt. Is this ok for you?

**Comment by zorggomat** [comment]
@umutcamliyurt hello! Please give me a little time so I can check the changes and build and test the project. Thank you for your participation. My email is xenomat@yandex.ru

File: FractalCryptGUI.pro
Comment (zorggomat) [inline FractalCryptGUI.pro:1]: Do we really need a second pro file?

File: LICENSE [hunk 1]
```diff
@@ -1,6 +1,6 @@
 MIT License
 
-Copyright (c) 2021 Ivan Serov
+Copyright (c) 2024 Nemesis
```
Comment (zorggomat) [inline LICENSE]: I think you should add rather than replace

File: core/aes.cpp [hunk 2]
```diff
@@ -1,13 +1,31 @@
 #include "aes.h"
+#include <QDebug>
+#include <openssl/rand.h>
+#include <argon2.h>
+#include <QRandomGenerator>
 
-AES::AES()
+AES::AES() : device(nullptr), pos(0), end(0), mode(Encrypt), success(false)
```
Comment (zorggomat) [inline core/aes.cpp]: It is more proper to initialize fields in the order they are defined
`AES::AES() : mode(Encrypt), device(nullptr), pos(0), end(0), success(false)`

File: core/aes.cpp [hunk 3]
```diff
@@ -26,109 +44,222 @@ void AES::setRange(qint64 pos, qint64 end)
...
+void AES::setPassword(const QString& password)
 {
-    if(password.isEmpty())
+    this->password = password;
+    if (password.isEmpty())
     {
         QRandomGenerator *random = QRandomGenerator::global();
         key.resize(32);
-        random->generate(key.begin(), key.end());
+        for (int i = 0; i < key.size(); ++i)
```
Comment (zorggomat) [inline core/aes.cpp]: For what reason did you choose to use the loop instead of calling a function with iterators?

File: core/aes.cpp [hunk 4]
```diff
@@ -26,109 +44,222 @@ void AES::setRange(qint64 pos, qint64 end)
...
-    if(!ctx) return false;
-    if(1 != EVP_EncryptInit_ex(ctx, EVP_aes_256_xts(), NULL, key, NULL)) return false;
+    if (!EVP_EncryptInit_ex(ctx, EVP_aes_256_xts(), NULL, reinterpret_cast<const unsigned char*>(key.constData()), iv)) {
+        qDebug() << "EVP_EncryptInit_ex failed.";
+        return false;
+    }
 
     unsigned char buffer[bufferSize];
-    unsigned char outBuffer[bufferSize];
+    unsigned char outBuffer[bufferSize + EVP_CIPHER_block_size(EVP_aes_256_xts())];
```
Comment (zorggomat) [inline core/aes.cpp]: I failed to build this
```
error C2131: expression did not evaluate to a constant
note: failure was caused by call of undefined function or one not declared 'constexpr'
note: see usage of 'EVP_aes_256_xts'
```

**Comment by zorggomat** [comment]
Hello @umutcamliyurt 
I wrote comments on your pull-request.
Also it would be great if you give me your email too.

**Comment by umutcamliyurt** [comment]
Thank you for your quick review :) My e-mail is: nemesisuks@protonmail.com

**Comment by umutcamliyurt** [comment]
My e-mail: nemesisuks@outlook.com

**Comment by umutcamliyurt** [comment]
Do my e-mails reach you? I think protonmail and outlook maybe blocked in russia

**Review by zorggomat — CHANGES_REQUESTED** [review]

File: FractalCryptGUI.pro
Comment (umutcamliyurt) [reply FractalCryptGUI.pro:1]: Deleted FractalCryptGUI.pro

File: LICENSE (hunk 1 above)
Comment (umutcamliyurt) [reply LICENSE]: Added it

File: FractalCryptGUI.pro
Comment (zorggomat) [inline FractalCryptGUI.pro:1]: File still exists

File: core/aes.cpp [hunk 5]
```diff
@@ -1,13 +1,31 @@
 #include "aes.h"
+#include <QDebug>
+#include <openssl/rand.h>
+#include <argon2.h>
```
Comment (zorggomat) [inline core/aes.cpp]: It is better to use `#include "argon2.h` because the header is located in the local directory

**Review by zorggomat — CHANGES_REQUESTED** [review]

File: core/aes.cpp (hunk 5 above)
Comment (umutcamliyurt) [reply core/aes.cpp]: fixed

File: FractalCryptGUI.pro
Comment (umutcamliyurt) [reply FractalCryptGUI.pro:1]: Deleted the file

**Pull request closed without merging**